### PR TITLE
Update typecheck err msg

### PR DESCRIPTION
**Pull request by hjtran**
The error message for applying a pcollection of incorrect type to a PTransform is really long and difficult to parse. For example, with this pipeline:
```
with beam.Pipeline() as p:                                                         
                                                                                   
  def square_root(x: float):                                                       
    return math.sqrt(x)                                                            
                                                                                   
  p | beam.Create(['foo', 'bar']) | beam.Map(square_root)  
  ```
  The error message is:
  ```
  output: Traceback (most recent call last):
  File "/opt/playground/backend/executable_files/7322872d-f8c2-4c0a-8ae7-b8f19386bc87/7322872d-f8c2-4c0a-8ae7-b8f19386bc87.py", line 48, in <module>
    p | beam.Create(['foo', 'bar']) | beam.Map(square_root)   
  File "/usr/local/lib/python3.10/site-packages/apache_beam/pvalue.py", line 138, in __or__
    return self.pipeline.apply(ptransform, self)
  File "/usr/local/lib/python3.10/site-packages/apache_beam/pipeline.py", line 746, in apply
    transform.type_check_inputs(pvalueish)
  File "/usr/local/lib/python3.10/site-packages/apache_beam/transforms/ptransform.py", line 949, in type_check_inputs
    raise TypeCheckError(
apache_beam.typehints.decorators.TypeCheckError: Type hint violation for 'Map(square_root)': requires <class 'float'> but got <class 'str'> for x
Full type hint:
IOTypeHints[inputs=((<class 'float'>,), {}), outputs=((Any,), {})]
strip_iterable()

based on:
  IOTypeHints[inputs=((<class 'float'>,), {}), outputs=((Iterable[Any],), {})]
  File "/opt/playground/backend/executable_files/7322872d-f8c2-4c0a-8ae7-b8f19386bc87/7322872d-f8c2-4c0a-8ae7-b8f19386bc87.py", line 48, in <module>
      p | beam.Create(['foo', 'bar']) | beam.Map(square_root)
  File "/usr/local/lib/python3.10/site-packages/apache_beam/transforms/core.py", line 2083, in Map
      wrapper = with_output_types(
  File "/usr/local/lib/python3.10/site-packages/apache_beam/typehints/decorators.py", line 861, in annotate_output_types
      f._type_hints = th.with_output_types(return_type_hint)  # pylint: disable=protected-access
  
  based on:
    IOTypeHints[inputs=((<class 'float'>,), {}), outputs=None]
    File "/opt/playground/backend/executable_files/7322872d-f8c2-4c0a-8ae7-b8f19386bc87/7322872d-f8c2-4c0a-8ae7-b8f19386bc87.py", line 48, in <module>
        p | beam.Create(['foo', 'bar']) | beam.Map(square_root)
    File "/usr/local/lib/python3.10/site-packages/apache_beam/transforms/core.py", line 2078, in Map
        wrapper = with_input_types(
    File "/usr/local/lib/python3.10/site-packages/apache_beam/typehints/decorators.py", line 774, in annotate_input_types
        th = getattr(f, '_type_hints', IOTypeHints.empty()).with_input_types(
```
This error message is really long and the extra "based on..." context (in my experience and the users I've talked to's experience) hasn't been helpful in figuring out the issue.

When there's a type hint issue, I want to know:
  - Which transform is causing the issue
  - What input type it expects
  - What input type it's getting instead
  - What transform is producing that input type.
  
  I've simplified the error message to answer those questions a bit more directly, so now the above example produces the following error message:
  ```
  apache_beam.typehints.decorators.TypeCheckError: The transform 'Map(square_root)' requires PCollections of type '<class 'float'>' but was applied to a PCollection of type '<class 'str'>' (produced by the transform 'Create').
  ```
  
  I've only applied this change to type hint violations for the main input, not for side inputs which I've left alone (though I think the "based on..." messages probably aren't useful there either)

### CODE REVIEW

**Comment by hjtran** [comment]
There are still some failing tests that are just checking the error string.

Before spending more time fixing them, I'd like to see what people think of this error message first.

**Comment by hjtran** [comment]
assign set of reviewers

**Comment by hjtran** [comment]
Bump @shunping @chamikaramj , could I get your quick feedback on just the error message here (no need to look at the code yet)? If it looks good, I'll fix up the rest of the tests and wait for another round of approval

**Comment by shunping** [comment]
@hjtran, I totally agree with you. I myself encountered a similar situation recently, and I think it is a great idea to simplify the error message and gives cx a better experience. Thanks for contributing!

+ @jrmccluskey, who is our expert of python typehint, for any more advice.

File: sdks/python/apache_beam/transforms/ptransform.py
Comment (jrmccluskey) [inline sdks/python/apache_beam/transforms/ptransform.py:954]: Is there a reason this type check is done up here rather than modifying the check nested in the for loop? It looks kind of redundant here.

File: sdks/python/pyproject.toml
Comment (jrmccluskey) [inline sdks/python/pyproject.toml:29]: Please revert this change (I'm assuming this was unintentional)

**Comment by jrmccluskey** [comment]
I definitely like the new form factor for the error, that's much cleaner. I've gotten decently good at parsing the current messages since I've been in and around the typehinting code a lot, but making it more clear for users is a big win (and may also make my life easier when I do more typehinting updates.)

File: sdks/python/apache_beam/transforms/ptransform.py
Comment (hjtran) [reply sdks/python/apache_beam/transforms/ptransform.py:954]: Yeah this is specifically a check for the main element while the loop is checking presumably side inputs.

File: sdks/python/pyproject.toml
Comment (hjtran) [reply sdks/python/pyproject.toml:29]: Huh, yeah I didn't change this intentionally. Maybe weird git operation artifact. Reverting

File: sdks/python/apache_beam/transforms/ptransform_test.py
Comment (hjtran) [inline sdks/python/apache_beam/transforms/ptransform_test.py:2138]: This and one of the other messages got too hard/frustrating to update to use with regex, so I just did some basic spot checking. The number of cases that the new asserts wouldn't catch but the old would I think are pretty few.

**Comment by hjtran** [comment]
Bump @jrmccluskey 
(note, I'm away from computer for the month of November after today)

**Review by jrmccluskey — APPROVED** [review]
LGTM, thank you!

**Pull request merged**
